ARSN-498: Test KMIP retry and cluster with pykmip

BourgoisMickael opened the pull request:
You can review by commit

Picked pykmip from cloudserver, updated it and test kmip with pykmip

BourgoisMickael committed:
ARSN-498: Test kmip retry on internal thales err

BourgoisMickael committed:
ARSN-498: Copy pykmip from cldsrv for real tests

BourgoisMickael committed:
ARSN-498: Functional test KMIP Cluster

bert-e commented:
# Hello bourgoismickael,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/arsenal/doc/user).


<details>
  <summary><b>Available options</b></summary>

  name   | description  | privileged   | authored
  ------ | ------------ | ------------ |----------
  `/after_pull_request` | Wait for the given pull request id to be merged before continuing with the current one. |  | 
  `/bypass_author_approval` | Bypass the pull request author's approval |  :star:  | 
  `/bypass_build_status` | Bypass the build and test status |  :star:  | 
  `/bypass_commit_size` | Bypass the check on the size of the changeset ```TBA``` |  :star:  | 
  `/bypass_incompatible_branch` | Bypass the check on the source branch prefix |  :star:  | 
  `/bypass_jira_check` | Bypass the Jira issue check |  :star:  | 
  `/bypass_peer_approval` | Bypass the pull request peers' approval |  :star:  | 
  `/bypass_leader_approval` | Bypass the pull request leaders' approval |  :star:  | 
  `/approve` | Instruct Bert-E that the author has approved the pull request. |  |  :writing_hand: 
  `/create_pull_requests` | Allow the creation of integration pull requests. |  | 
  `/create_integration_branches` | Allow the creation of integration branches. |  | 
  `/no_octopus` | Prevent Wall-E from doing any octopus merge and use multiple consecutive merge instead |  | 
  `/unanimity` | Change review acceptance criteria from `one reviewer at least` to `all reviewers` |  | 
  `/wait` | Instruct Bert-E not to run until further notice. |  | 
  

</details>



<details>
  <summary><b>Available commands</b></summary>

  name   | description  | privileged
  ------ | ------------ | ------------
  `/help` | Print Bert-E's manual in the pull request. | 
  `/status` | Print Bert-E's current status in the pull request ```TBA``` | 
  `/clear` | Remove all comments from Bert-E from the history ```TBA``` | 
  `/retry` | Re-start a fresh build ```TBA``` | 
  `/build` | Re-start a fresh build ```TBA``` | 
  `/force_reset` | Delete integration branches & pull requests, and restart merge process from the beginning. | 
  `/reset` | Try to remove integration branches unless there are commits on them which do not appear on the source branch. | 
  
</details>




*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue ARSN-498 contains:


* `7.70.46`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `7.70.46`

* `8.1.157`

* `8.2.19`


Please check the `Fix Version/s` of ARSN-498, or the target
branch of this pull request.

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue ARSN-498 contains:


* `7.70.46`

* `8.1.157`

* `8.2.18`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `7.70.46`

* `8.1.157`

* `8.2.19`


Please check the `Fix Version/s` of ARSN-498, or the target
branch of this pull request.

bert-e commented:
# Request integration branches



Waiting for integration branch creation to be requested by the user.

To request integration branches, please comment on this pull request with the following command:

```
/create_integration_branches
```

Alternatively, the `/approve` and `/create_pull_requests` commands will automatically
create the integration branches.

BourgoisMickael committed:
ARSN-498: Update pykmip for fixes and better perf

BourgoisMickael committed:
ARSN-498: Add a kmip stop for graceful shutdown

And replace end event with close event for socket
without tls session

For tests not to keep a lot of connections open

BourgoisMickael committed:
ARSN-498: Add SAN in kmip cert for kmip cluster

Allow DNS: *.pykmip.local, localhost, IP: 127.0.0.[1-50]

Addition in certificate:

```
+         X509v3 extensions:
+             X509v3 Subject Alternative Name:
+                 DNS:pykmip.local, DNS:*.pykmip.local, DNS:localhost, IP Address:127.0.0.1, IP Address:127.0.0.2, IP Address:127.0.0.3, IP Address:127.0.0.4, IP Address:127.0.0.5, IP Address:127.0.0.6, IP Address:127.0.0.7, IP Address:127.0.0.8, IP Address:127.0.0.9, IP Address:127.0.0.10, IP Address:127.0.0.11, IP Address:127.0.0.12, IP Address:127.0.0.13, IP Address:127.0.0.14, IP Address:127.0.0.15, IP Address:127.0.0.16, IP Address:127.0.0.17, IP Address:127.0.0.18, IP Address:127.0.0.19, IP Address:127.0.0.20, IP Address:127.0.0.21, IP Address:127.0.0.22, IP Address:127.0.0.23, IP Address:127.0.0.24, IP Address:127.0.0.25, IP Address:127.0.0.26, IP Address:127.0.0.27, IP Address:127.0.0.28, IP Address:127.0.0.29, IP Address:127.0.0.30, IP Address:127.0.0.31, IP Address:127.0.0.32, IP Address:127.0.0.33, IP Address:127.0.0.34, IP Address:127.0.0.35, IP Address:127.0.0.36, IP Address:127.0.0.37, IP Address:127.0.0.38, IP Address:127.0.0.39, IP Address:127.0.0.40, IP Address:127.0.0.41, IP Address:127.0.0.42, IP Address:127.0.0.43, IP Address:127.0.0.44, IP Address:127.0.0.45, IP Address:127.0.0.46, IP Address:127.0.0.47, IP Address:127.0.0.48, IP Address:127.0.0.49, IP Address:127.0.0.50
```

Serial Number, Validity date and signature updated as well

BourgoisMickael commented:
/create_integration_branches

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/8.1/improvement/ARSN-498-test-kmip` with contents from `improvement/ARSN-498-test-kmip`
and `development/8.1`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/8.1/improvement/ARSN-498-test-kmip origin/development/8.1
 $ git merge origin/improvement/ARSN-498-test-kmip
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/8.1/improvement/ARSN-498-test-kmip
```





*The following options are set:* **create_integration_branches**

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/8.2/improvement/ARSN-498-test-kmip` with contents from `w/8.1/improvement/ARSN-498-test-kmip`
and `development/8.2`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/8.2/improvement/ARSN-498-test-kmip origin/development/8.2
 $ git merge origin/w/8.1/improvement/ARSN-498-test-kmip
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/8.2/improvement/ARSN-498-test-kmip
```





*The following options are set:* **create_integration_branches**

bert-e commented:
# History mismatch



Merge commit #bd9df9f866772cb64f82e34b2afb14e908026a3a on the integration branch
`w/8.1/improvement/ARSN-498-test-kmip` is merging a branch which is neither the current
branch `improvement/ARSN-498-test-kmip` nor the development branch
`development/8.1`.

It is likely due to a rebase of the branch `improvement/ARSN-498-test-kmip` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **create_integration_branches**

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/8.2/improvement/ARSN-498-test-kmip` with contents from `w/8.1/improvement/ARSN-498-test-kmip`
and `development/8.2`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/8.2/improvement/ARSN-498-test-kmip origin/development/8.2
 $ git merge origin/w/8.1/improvement/ARSN-498-test-kmip
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/8.2/improvement/ARSN-498-test-kmip
```





*The following options are set:* **create_integration_branches**

bert-e commented:
# History mismatch



Merge commit #fa0f2006053f9c27ece10d726885ff52cf472116 on the integration branch
`w/8.1/improvement/ARSN-498-test-kmip` is merging a branch which is neither the current
branch `improvement/ARSN-498-test-kmip` nor the development branch
`development/8.1`.

It is likely due to a rebase of the branch `improvement/ARSN-498-test-kmip` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **create_integration_branches**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers











*The following options are set:* **create_integration_branches**

BourgoisMickael commented on lib/network/kmip/transport/TransportTemplate.ts:
```diff
@@ -221,11 +223,16 @@ export default class TransportTemplate {
+     * @param onEndCallback - callback called once the socket is ended and pipelines are drained
      */
-    end() {
+    end(onEndCallback?: any) {
```
Yes it is actually called...

tcarmet commented on .github/workflows/tests-kmip.yaml:
```diff
@@ -0,0 +1,49 @@
+    - .github/docker-compose.yaml
+    - .github/pykmip/**
+    - lib/network/kmip/**
+    - tests/functional/kmip/**
```
`paths` does not work as you would expect with `push` when working on the feature branches, and does not work well with gitwaterflow integration branches as well (`w/**`).

As a quick fix I can recommend you to use a combination of `push` and `pull_request` trigger, even though it won't be perfect, it will work better:

```suggestion
  pull_request:
    branches:
    - development/*
    - hotfix/*
    paths:
    - .github/workflows/tests-kmip.yaml
    - .github/docker-compose.yaml
    - .github/pykmip/**
    - lib/network/kmip/**
    - tests/functional/kmip/**
  push:
    branches:
    - q/*
    - w/**
    paths:
    - .github/workflows/tests-kmip.yaml
    - .github/docker-compose.yaml
    - .github/pykmip/**
    - lib/network/kmip/**
    - tests/functional/kmip/**
```

For a more accurate way I recommend you checking out the use of https://github.com/tj-actions/changed-files/, we do it in Artesca for the documentation build, if you want to see a real life example of it. 

But honestly, if it doesn't take too long, I would just leave it. I see no issue in having it around:

```suggestion
  push:
    branches-ignore:
    - 'development/**'
```

BourgoisMickael replied:
let's try with both push and pull_request

tcarmet commented on .github/docker-compose.yaml:
```diff
@@ -0,0 +1,13 @@
+  pykmip:
+    network_mode: "host"
+    profiles: ['pykmip']
+    #image: ${PYKMIP_IMAGE:-ghcr.io/scality/cloudserver/pykmip}
```
```suggestion
```

tcarmet commented on .github/pykmip/Dockerfile:
```diff
@@ -0,0 +1,29 @@
+        build-base \
+        curl \
+        git && \
+    git clone https://github.com/openkmip/pykmip.git && \
```
Was gonna say to specify a version here, but I see that we havent got releases in a while and there has been some modifications to the repo a while ago. Is that why you're not specifying a tag here?

BourgoisMickael replied:
yep I want the latest changes of the repo to include any latest fix as it almost never receive any updates.

But maybe I should lock on the commit hash?

tcarmet replied:
Yeah, all things considered (the repo not having much updates and all) either is fine I think. Mostly wanted to make sure this was on purpose.

anurag4DSB commented on tests/functional/kmip/highlevel.spec.js:
```diff
@@ -120,4 +122,130 @@ describe('KMIP High Level Driver', () => {
+        const message = '[NCERRUnauthorizedAccess]: Token has been revoked';
+
+        // the next call should return a thales session revoked error only once
+        const stub = sinon.stub(kmipClient.kmip.transport.channel, '_transform')
```
I see that there is no cleanup, I'd recommend cleaning up spies/stubs via `afterEach()` to reduce test flakiness if extended or parallelized.
Something like

```
afterEach(() => {
  sinon.restore();
});
```

anurag4DSB commented on tests/functional/kmip/highlevel.spec.js:
```diff
@@ -120,4 +122,130 @@ describe('KMIP High Level Driver', () => {
+            `tls write should have been called twice because of retry instead of ${tlsWriteSpy.callCount}`);
+        assert.strictEqual(Buffer.compare(tlsWriteSpy.getCall(0).args[0], tlsWriteSpy.getCall(1).args[0]), 0,
+            'tls write should have twice the same argument');
+        assert.strictEqual(stub.callCount, 2,
```
Do we plan to increase this count in the future?

BourgoisMickael replied:
No, at this place we don't expect any configurable retry. It's only 1 specific message we know we can retry immediately otherwise any other error message should be sent back to client

anurag4DSB commented on tests/functional/kmip/highlevel.spec.js:
```diff
@@ -120,4 +122,130 @@ describe('KMIP High Level Driver', () => {
+
+        assert.strictEqual(tlsWriteSpy.callCount, 2,
+            `tls write should have been called twice because of retry instead of ${tlsWriteSpy.callCount}`);
+        assert.strictEqual(Buffer.compare(tlsWriteSpy.getCall(0).args[0], tlsWriteSpy.getCall(1).args[0]), 0,
```
Shouldn't retry logic should preserve the original payload for idempotency

BourgoisMickael replied:
Yes, this is was test checks

> Compares buf with target and returns a number indicating whether buf comes before, after, or is the same as target in sort order. Comparison is based on the actual sequence of bytes in each Buffer.
>
> 0 is returned if target is the same as buf
> 1 is returned if target should come before buf when sorted.
> -1 is returned if target should come after buf when sorted.


https://nodejs.org/api/buffer.html#bufcomparetarget-targetstart-targetend-sourcestart-sourceend

BourgoisMickael replied:
Or you mean we should check the buffer is actually the same instance instead of checking the content of the buffer ?

anurag4DSB replied:
No, not the same instance, I was referring to buffer content equality, not reference equality. The current check using `Buffer.compare(...) === 0` is exactly what I was expecting: validating that the payload is byte-for-byte identical on retry to ensure idempotency.

So all good here, just wanted to confirm that this behavior was intentional. 🙌

As a side note: if we’re targeting a modern Node.js version, [Buffer.equals()](https://nodejs.org/api/buffer.html#bufequalsotherbuffer) might make the intent a bit more readable here, semantically expresses equality instead of comparing sort order.

anurag4DSB commented on tests/functional/kmip/highlevel.spec.js:
```diff
@@ -120,4 +122,130 @@ describe('KMIP High Level Driver', () => {
+        const tlsConnectSpy = sinon.spy(kmipClient.kmip.transport.channel, 'connect');
+
+
+        const encrypted2 = await util.promisify(kmipClient.cipherDataKey.bind(kmipClient))(
```
I would rename this from `encrypted2` to `retriedEncrypted` to better reflect intent

anurag4DSB commented on tests/functional/kmip/highlevel.spec.js:
```diff
@@ -120,4 +122,130 @@ describe('KMIP High Level Driver', () => {
         });
     });
+
+    it('should retry without kmipHandshake on thales session revoked', async () => {
```
Both tests use nearly identical setup (options, client init, healthcheck). could we factor out common setup into a shared helper or `beforeEach()` block to improve readability and reduce duplication. Example: `initTestClient()`.

BourgoisMickael replied:
Fixed this file in [945dbac](https://github.com/scality/Arsenal/pull/2402/commits/945dbac1e9f510b39a7b9e61643a3bab2eae4cc8) and [b11e602](https://github.com/scality/Arsenal/pull/2402/commits/b11e6020627cf0321681b1d348a7194e0a3ec701)

anurag4DSB commented on tests/functional/pykmip/ClusterClient.spec.js:
```diff
@@ -0,0 +1,508 @@
+    let kmsKey;
+
+    // Make sure to always disconnect any client
+    afterEach(async () => {
```
Can we factor out repeated client teardown + key deletion into a shared helper to keep the `afterEach` blocks DRY and easier to maintain.

BourgoisMickael replied:
Refactored this file here [716b69c](https://github.com/scality/Arsenal/pull/2402/commits/716b69cb3f7a901b88a948d1ea39d5f434de5d80)

anurag4DSB commented on tests/functional/pykmip/ClusterClient.spec.js:
```diff
@@ -0,0 +1,508 @@
+        promises = promisify(client);
+    });
+
+    afterEach(async () => {
```
We should add `sinon.restore()` here as well, like in the cluster tests, to ensure future added spies don't leak between test cases.

anurag4DSB commented on tests/functional/pykmip/ClusterClient.spec.js:
```diff
@@ -0,0 +1,508 @@
+    logger,
+} = require('../../utils/kmip/ersatz');
+
+jest.setTimeout(5000);
```
Please consider making the timeout configurable or environment-dependent.
In CI under load, 5s might not be enough and could cause flaky tests.

BourgoisMickael replied:
This is configured for CI, it runs alone in a separate test CI for kmip with pykmip.

I tested with many more hosts in cluster than 10 and more encryption than 100.

5s is way more than enough for a cluster of 10 hosts

anurag4DSB replied:
Why not add it in package.json then?

BourgoisMickael replied:
The test is not in package.json, it's executed directly in its own CI workflow `yarn jest tests/functional/pykmip/ClusterClient.spec.js` separate from the other jest tests

So there is no other config to add, everything is self contained for this test

anurag4DSB replied:
In that case we should add it to package.json scripts as thats expected for our convention.

BourgoisMickael replied:
added a script in package.json for pykmip tests

anurag4DSB commented on tests/functional/pykmip/ClusterClient.spec.js:
```diff
@@ -0,0 +1,508 @@
+            deciphered = await promises.decipherDataKey(1, kmsKey, ciphered, logger);
+        }, showArsenalErr);
+
+        assert.ok(Buffer.compare(dataKey, deciphered) === 0,
```
`Buffer.equals` if supported, could help for better readability.

anurag4DSB commented on tests/functional/pykmip/ClusterClient.spec.js:
```diff
@@ -0,0 +1,508 @@
+                { tls: Object.assign({}, transport.tls, { host: `127.0.0.${i + 1}` }) }),
+        );
+        transports.push(Object.assign({}, transport,
+            // not in certificate, will fail connection
```
This works for simulating host failure, but please consider being explicit:
“Host not in TLS cert, used to simulate healthcheck failure and trigger fallback logic.”

anurag4DSB commented on tests/functional/pykmip/ClusterClient.spec.js:
```diff
@@ -0,0 +1,508 @@
+            clock = sinon.useFakeTimers({ shouldAdvanceTime: true });
+        });
+
+        afterEach(async () => {
+            if (clock) {
+                clock.reset();
+            }
+        });
```
What about using clock.restore() after reset to fully clean up timers and restore global state.
Just resetting time might leave patched globals active.

BourgoisMickael replied:
yep with `restore` I can skip the `reset` as the stub is removed

anurag4DSB commented on tests/functional/pykmip/ClusterClient.spec.js:
```diff
@@ -0,0 +1,508 @@
+
+            spys.cipherDataKeys.forEach(spy => spy.resetHistory());
+
+            await stopSocat(socatProcess);
```
Consider wrapping stopSocat in try/finally or putting it in `afterEach` to avoid port leakage in case the test fails midway. Helpful in CI stability.

BourgoisMickael replied:
Indeed forgot to include it in the `afterEach`

anurag4DSB commented on tests/functional/pykmip/ClusterClient.spec.js:
```diff
@@ -0,0 +1,508 @@
+            await stopSocat(socatProcess);
+
+            // extract unhealthy spy from healthy
+            const spysCipherUnhealthy = spys.cipherDataKeys.splice(spys.cipherDataKeys.length - 1, 1);
```
Is it always true that always last client is the one that fails

BourgoisMickael replied:
in that test yes because I pushed the client that I use to simulate disconnection at the end of the transport list

BourgoisMickael committed:
ARSN-498: Factorize kmip test and afterEach

bert-e commented:
# History mismatch



Merge commit #7e78f85ee5b506e41b3953493be00e79302cbb60 on the integration branch
`w/8.1/improvement/ARSN-498-test-kmip` is merging a branch which is neither the current
branch `improvement/ARSN-498-test-kmip` nor the development branch
`development/8.1`.

It is likely due to a rebase of the branch `improvement/ARSN-498-test-kmip` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **create_integration_branches**

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **create_integration_branches**

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/8.1/improvement/ARSN-498-test-kmip` with contents from `improvement/ARSN-498-test-kmip`
and `development/8.1`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/8.1/improvement/ARSN-498-test-kmip origin/development/8.1
 $ git merge origin/improvement/ARSN-498-test-kmip
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/8.1/improvement/ARSN-498-test-kmip
```





*The following options are set:* **create_integration_branches**

BourgoisMickael commented:
/approve

bert-e commented:
# Build failed



The [build](https://github.com/scality/Arsenal/actions?query=branch%3Aw/8.2/improvement/ARSN-498-test-kmip) for [commit](https://github.com/scality/Arsenal/commit/b475f41914385f5f62f2bf0a546b8767118d1d5e) did not succeed in branch [w/8.2/improvement/ARSN-498-test-kmip](https://github.com/scality/arsenal/compare/development/8.2...w/8.2/improvement/ARSN-498-test-kmip)



*The following options are set:* **approve, create_integration_branches**

bert-e commented:
# Build failed



The [build](https://github.com/scality/Arsenal/actions?query=branch%3Aw/8.2/improvement/ARSN-498-test-kmip) for [commit](https://github.com/scality/Arsenal/commit/cbb4ff388b544b486d8826ccad5d0c03d63552fb) did not succeed in branch [w/8.2/improvement/ARSN-498-test-kmip](https://github.com/scality/arsenal/compare/development/8.2...w/8.2/improvement/ARSN-498-test-kmip)



*The following options are set:* **approve, create_integration_branches**

BourgoisMickael committed:
ARSN-498: Node16 supports es2021

https://node.green/#ES2021

Needed for AggregateError

BourgoisMickael committed:
ARSN-498: Handle dual-stack AggregateError

Found in dev/8 branch on node22 on kmip with default host localhost

The error message is more clear:
```diff
- AggregateError
+ AggregateError: Error: connect ECONNREFUSED ::1:5696, Error: connect ECONNREFUSED 127.0.0.1:5696
```

The test produce clearer diff as well in case of error instead of boolean

BourgoisMickael committed:
ARSN-498: Hide Kmip tls internal error message

bert-e commented:
# Conflict




A conflict has been raised during the update of
integration branch `w/8.1/improvement/ARSN-498-test-kmip` with contents from `improvement/ARSN-498-test-kmip`
and `development/8.1`.

Please resolve the conflict on **the integration branch** (`w/8.1/improvement/ARSN-498-test-kmip`).


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout w/8.1/improvement/ARSN-498-test-kmip
 $ git pull  # or "git reset --hard origin/w/8.1/improvement/ARSN-498-test-kmip"
 $ git merge origin/development/8.1
 $ # <intense conflict resolution>
 $ git commit
 $ git merge origin/improvement/ARSN-498-test-kmip
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/8.1/improvement/ARSN-498-test-kmip
```





*The following options are set:* **create_integration_branches**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers











*The following options are set:* **create_integration_branches**

anurag4DSB reviewed (approved):
If there is any change in integration branches, please do message me on Slack, I would like to review them too.

BourgoisMickael commented:
/approve

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/7.70`

* :heavy_check_mark: `development/8.1`

* :heavy_check_mark: `development/8.2`



The following branches have **NOT** changed:

* `development/6.4`
* `development/7.10`
* `development/7.4`



Please check the status of the associated issue ARSN-498.

Goodbye bourgoismickael.




*The following options are set:* **approve, create_integration_branches**
